File: src/main/java/com/example/expression/ExpressionParser.java
package com.example.expression;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Парсер и вычислитель математических выражений.
 * <p>
 * Объединяет токенизатор, парсер Пратта и вычислитель RPN для полного
 * цикла обработки математических выражений.
 * </p>
 * 
 * <h3>Архитектура (аналогична JS-версии):</h3>
 * <ol>
 * <li><b>ExpressionTokenizer</b> - разбивает строку на токены</li>
 * <li><b>PrattParser</b> - преобразует токены в RPN</li>
 * <li><b>ExpressionEvaluator</b> - вычисляет RPN</li>
 * </ol>
 * 
 * <h3>Пример использования:</h3>
 * <pre>
 * ExpressionParser parser = new ExpressionParser("2 + 3 * sin(pi / 2)");
 * double result = parser.evaluate();  // 5.0
 * </pre>
 */
public class ExpressionParser {
    
    private final ExpressionEvaluator evaluator;
    private final List<Token> rpn;
    private final Map<String, Double> variables;
    
    /**
     * Создаёт парсер для указанного выражения.
     * 
     * @param expression математическое выражение
     * @throws IllegalArgumentException при ошибках синтаксиса
     */
    public ExpressionParser(String expression) {
        // Инициализируем компоненты из единого реестра
        Map<String, Double> constants = ExpressionDefinitions.CONSTANTS;
        Map<String, ExpressionEvaluator.FunctionDef> functions = ExpressionDefinitions.FUNCTIONS;
        Map<String, ExpressionEvaluator.OperatorDef> operators = ExpressionDefinitions.OPERATORS;
        
        // Создаём токенизатор
        ExpressionTokenizer tokenizer = new ExpressionTokenizer(
            ExpressionDefinitions.FUNCTION_NAMES,
            ExpressionDefinitions.CONSTANT_NAMES
        );
        
        // Создаём парсер
        PrattParser parser = new PrattParser(functions);
        
        // Создаём вычислитель
        this.evaluator = new ExpressionEvaluator(constants, functions, operators);
        
        // Парсим выражение
        List<Token> tokens = tokenizer.tokenize(expression);
        this.rpn = parser.parse(tokens);
        
        // Инициализируем переменные
        this.variables = new HashMap<>();
    }
    
    /**
     * Устанавливает значение переменной.
     * 
     * @param name имя переменной
     * @param value значение
     */
    public void setVariable(String name, double value) {
        variables.put(name, value);
    }
    
    /**
     * Вычисляет выражение.
     * 
     * @return результат вычисления
     * @throws IllegalArgumentException при ошибках вычисления
     */
    public double evaluate() {
        return evaluator.evaluate(rpn, variables);
    }
    
    /**
     * Возвращает RPN для отладки.
     */
    public List<Token> getRPN() {
        return rpn;
    }
    
}

File: js/expression_parser.js
class ExpressionParser {
    constructor(expression) {
        const functions = this.initFunctions()
        const constants = this.initConstans()
        const operators = this.initOperators()

        const tokenizer = new ExpressionTokenizer({
            functions: Object.keys(functions),
            constants: Object.keys(constants)
        })

        const parser = new PrattParser(functions)

        this.evaluator = new ExpressionEvaluator({constants, functions, operators})
        this.rpn = parser.parse(tokenizer.tokenize(expression))
        this.variables = {}
    }

    setVariable(name, value) {
        this.variables[name] = value
    }

    evaluate() {
        return this.evaluator.evaluate({rpn: this.rpn, variables: this.variables})
    }

    initFunctions() {        
        return {
            "sin": {args: 1, evaluate: Math.sin},
            "cos": {args: 1, evaluate: Math.cos},
            "tan": {args: 1, evaluate: Math.tan},
            "max": {args: 2, evaluate: Math.max},
        }
    }

    initConstans() {
        return {
            "pi": Math.PI,
            "e": Math.E
        }
    }

    initOperators() {
        return {
            "+": {args: 2, evaluate: (arg1, arg2) => arg1 + arg2},
            "-": {args: 2, evaluate: (arg1, arg2) => arg1 - arg2},
            "*": {args: 2, evaluate: (arg1, arg2) => arg1 * arg2},
            "/": {args: 2, evaluate: (arg1, arg2) => arg1 / arg2},
            "^": {args: 2, evaluate: (arg1, arg2) => arg1 ** arg2},
            "~": {args: 1, evaluate: (arg) => -arg}
        }
    }
}
